src/extension-runners/firefox-desktop.js

```diff
@@ -0,0 +1,229 @@
+    await this.startFirefoxInstance();
+  }
+
+  async reloadAllExtensions(): Promise<void> {
+    const reloadErrors = new Map();
+    for (const {sourceDir} of this.params.extensions) {
+      try {
+        await this.reloadExtensionBySourceDir(sourceDir);
+      } catch (error) {
+        reloadErrors.set(sourceDir, error);
+      }
+    }
+
+    if (reloadErrors.size > 0) {
+      return Promise.reject(new MultipleExtensionsReloadError(reloadErrors));
+    }
+  }
+
+  async reloadExtensionBySourceDir(extensionSourceDir: string): Promise<void> {
+    const addonId = this.reloadableExtensions.get(extensionSourceDir);
+    if (!addonId) {
+      throw new WebExtError('Extension not reloadable');
+    }
+
+    await this.remoteFirefox.reloadAddon(addonId);
+  }
+
+  registerCleanup(fn: Function): void {
+    this.cleanupCallbacks.add(fn);
+  }
+
+  async exit(): Promise<void> {
+    if (!this.runningInfo || !this.runningInfo.firefox) {
+      throw new WebExtError('No firefox instance is currently running');
+    }
+
+    this.runningInfo.firefox.kill();
+  }
+
+  // Private methods.
+
+  async setupProfileDir() {
+    const {
+      customPrefs,
+      extensions,
+      keepProfileChanges,
+      preInstall,
+      profilePath,
+      firefoxApp,
+    } = this.params;
+
+    if (profilePath) {
+      if (keepProfileChanges) {
+        log.debug(`Using Firefox profile from ${profilePath}`);
+        this.profile = await firefoxApp.useProfile(profilePath, {customPrefs});
+      } else {
+        log.debug(`Copying Firefox profile from ${profilePath}`);
+        this.profile = await firefoxApp.copyProfile(profilePath, {customPrefs});
+      }
+    } else {
+      log.debug('Creating new Firefox profile');
+      this.profile = await firefoxApp.createProfile({customPrefs});
+    }
+
+    // preInstall the extensions if needed.
+    if (preInstall) {
+      for (const extension of extensions) {
+        await firefoxApp.installExtension({
+          asProxy: true,
+          extensionPath: extension.sourceDir,
+          manifestData: extension.manifestData,
+          profile: this.profile,
+        });
+      }
+    }
+  }
+
+  async startFirefoxInstance() {
+    const {
+      browserConsole,
+      extensions,
+      firefoxBinary,
+      preInstall,
+      startUrl,
+      firefoxApp,
+      firefoxClient,
+    } = this.params;
+
+    const binaryArgs = [];
+
+    if (browserConsole) {
+      binaryArgs.push('-jsconsole');
+    }
+    if (startUrl) {
+      const urls = Array.isArray(startUrl) ? startUrl : [startUrl];
+      for (const url of urls) {
+        binaryArgs.push('--url', url);
+      }
+    }
+
+    this.runningInfo = await firefoxApp.run(this.profile, {
+      firefoxBinary, binaryArgs,
+    });
+
+    this.runningInfo.firefox.on('close', () => {
+      for (const fn of this.cleanupCallbacks) {
+        try {
+          fn();
+        } catch (error) {
+          log.debug(`Exception on executing cleanup callback: ${error}`);
```

Comment:
Should this be `log.error()` instead? Would it be typical for a cleanup function to throw an error that one can ignore?